### PR TITLE
[BEAM-65] Splittable DoFn: Gives the runner access to RestrictionTracker

### PR DESCRIPTION
**Pull request by jkff**
Changes the SplittableParDo transform so that ProcessFn uses a
runner-supplied hook to run the @ProcessElement method, giving
it, among other things, the RestrictionTracker, so the runner
can initiate checkpointing/splitting with it at will.

Introduces a default implementation of said hook, which limits
the number of outputs and duration of the call. This implementation
is used in tests and in Direct runner. Dataflow Streaming runner
will also use this implementation, while Dataflow Batch runner
will be more sophisticated.

R: @kennknowles

### CODE REVIEW

**Comment by jkff** [comment]
Oops, forgot to push a file. Fixed.

**Comment by jkff** [comment]
Actually hold on with reviewing this for a bit while I understand whether this is really the best way to abstract this, by looking at Dataflow worker.

**Comment by jkff** [comment]
retest this please

**Comment by jkff** [comment]
Definitely ready for review. I have a prototype in Dataflow runner doesn't run yet, but demonstrates that this is the right API.

**Comment by jkff** [comment]
Ping.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/OutputAndTimeBoundedSplittableProcessElementInvoker.java [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    }
+
+    private void noteOutput() {
+      ++numOutputs;
+      if (numOutputs >= maxNumOutputs) {
+        initiateCheckpoint();
+      }
+    }
+
+    private synchronized void initiateCheckpoint() {
```
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/OutputAndTimeBoundedSplittableProcessElementInvoker.java]: It seems the purpose here is to make sure only one thread ever enters the (presumably-expensive) act of `tracker.checkpoint()`.

To embed the access policy and use futures more idiomatically, I think you could give `checkpoint` the type `AtomicReference<Future<CheckpointT>>`.

Failing that, give it a getter and a setter that are both synchronized, put them next to each other, and call them out loudly on the field declaration.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/OutputAndTimeBoundedSplittableProcessElementInvoker.java [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+  @Override
+  public Result invokeProcessElement(
+      DoFnInvoker<InputT, OutputT> invoker,
+      final WindowedValue<InputT> element,
+      final TrackerT tracker) {
+    final ProcessContext processContext = new ProcessContext(element, tracker);
+    DoFn.ProcessContinuation cont =
+        invoker.invokeProcessElement(
+            new DoFnInvoker.FakeArgumentProvider<InputT, OutputT>() {
```
Comment (jkff) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/OutputAndTimeBoundedSplittableProcessElementInvoker.java]: Done, put in all the methods explicitly. A separate PR removes the annoying inputProvider/outputReceiver methods.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/OutputAndTimeBoundedSplittableProcessElementInvoker.java (hunk 1 above)
Comment (jkff) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/OutputAndTimeBoundedSplittableProcessElementInvoker.java]: I couldn't come up with cleaner code for this, but I moved the accesses to checkpoint closer together, and added a bunch of comments.

**Comment by jkff** [comment]
PTAL.

**Comment by dhalperi** [comment]
What JIRA is this under? What's the rationale for this change?

**Comment by kennknowles** [comment]
Yea this should be tagged with [BEAM-65](https://issues.apache.org/jira/browse/BEAM-65).

**Comment by jkff** [comment]
@dhalperi Tagged with JIRA; the rationale for this change is, I believe, explained in the commit message.

**Comment by dhalperi** [comment]
+1 thanks Eugene!

**Comment by jkff** [comment]
Added tests (while I was at it, moved OffsetRange and its Tracker to SDK since that'd be the third time I'd use it, including some pending code of mine where I use it in an integration test).

**Comment by jkff** [comment]
retest this please

**Comment by jkff** [comment]
Coveralls is weird. It says SplittableProcessElementsEvaluatorFactory is basically not covered, but it is covered by SplittableParDo tests in direct runner.

**Comment by kennknowles** [comment]
This is a Jacoco configuration issue. All coveralls does is present the data, which is essentially available. I won't block anything on Coveralls.

To easily browse yourself, run `mvn jacoco:report`

**Comment by dhalperi** [comment]
@jkff -- I believe this is induced by our crazy test setup. The SDK tests are covered in the direct runner module, etc. The issue is that we have not asked jacoco to combine these things into one report.

I spent a few minutes trying to fix this using newer versions of plugins that claim to be able to do it, but stalled: https://github.com/apache/beam/pull/1800

**Comment by kennknowles** [comment]
The Java code LGTM. Seems to have pulled in a ton of Python stuff.

**Comment by kennknowles** [comment]
I'm going to get green, fire off a postcommit, then merge. If  you'd like to fix up to preserve merge coherence you can, or I will squash when I merge, losing the history.

**Comment by jkff** [comment]
Squashing is fine.

**Comment by kennknowles** [comment]
https://builds.apache.org/view/Beam/job/beam_PostCommit_Java_RunnableOnService_Dataflow/2167/
https://builds.apache.org/view/Beam/job/beam_PostCommit_Java_RunnableOnService_Flink/1506/
https://builds.apache.org/view/Beam/job/beam_PostCommit_Java_RunnableOnService_Spark/816/
https://builds.apache.org/view/Beam/job/beam_PostCommit_Java_RunnableOnService_Apex/372/

**Pull request closed without merging**